inject/inject.vala

```diff
@@ -0,0 +1,421 @@
+namespace Frida.Inject {
+	private static Application application;
+
+	private static bool output_version;
+	private static bool disable_jit;
+	private static bool enable_development;
+	private static int pid;
+	private static string script_file;
+
+	static const OptionEntry[] options = {
+		{ "version", 0, 0, OptionArg.NONE, ref output_version, "Output version information and exit", null },
+		{ "disable-jit", 0, 0, OptionArg.NONE, ref disable_jit, "Disable the JIT runtime", null },
+		{ "development", 'D', 0, OptionArg.NONE, ref enable_development, "Enable development mode", null },
+		{ "pid", 'p', 0, OptionArg.INT, ref pid, null, "PID" },
+		{ "script", 's', 0, OptionArg.FILENAME, ref script_file, null, "JAVASCRIPT_FILENAME" },
+		{ null }
+	};
+
+	private static int main (string[] args) {
+#if !WINDOWS
+		Posix.setsid ();
+#endif
+
+		Environment.init ();
+
+		pid = -1;
+		script_file = "";
+
+		try {
+			var ctx = new OptionContext ();
+			ctx.set_help_enabled (true);
+			ctx.add_main_entries (options, null);
+			ctx.parse (ref args);
+
+			if (output_version) {
+				stdout.printf ("%s\n", version_string ());
+				return 0;
+			}
+		} catch (OptionError e) {
+			stdout.printf ("%s\n", e.message);
+			stdout.printf ("Run '%s --help' to see a full list of available command line options.\n", args[0]);
+			return 1;
+		}
+
+		if (pid == -1) {
+			stdout.printf ("pid must be specified\n");
+			return 1;
+		}
+
+		if (script_file == "") {
+			stdout.printf ("javascript filename must be specified\n");
+			return 1;
+		}
+
+		application = new Application ();
+
+#if !WINDOWS
+		Posix.signal (Posix.SIGINT, (sig) => {
+			application.shutdown ();
+		});
+		Posix.signal (Posix.SIGTERM, (sig) => {
+			application.shutdown ();
+		});
+#endif
+
+		try {
+			application.run (pid, script_file, disable_jit, enable_development);
+		} catch (Error e) {
+			printerr ("Unable to start inject: %s\n", e.message);
+			return 1;
+		}
+
+		application = null;
+
+		return 0;
+	}
+
+	namespace Environment {
+		public extern void init ();
+	}
+
+	public class Application : Object {
+		DeviceManager device_manager;
+		private int pid;
+		private string script_file;
+		private bool disable_jit;
+		private bool enable_development;
+		private ScriptRunner script_runner;
+
+		private MainLoop loop;
+		private bool stopping;
+
+		construct {
+		}
+
+		public void run (int pid, string script_file, bool disable_jit, bool enable_development) throws Error {
+			this.pid = pid;
+			this.script_file = script_file;
+			this.disable_jit = disable_jit;
+			this.enable_development = enable_development;
+
+			Idle.add (() => {
+				start.begin ();
+				return false;
+			});
+
+			loop = new MainLoop ();
+			loop.run ();
+		}
+		
+		public async void start () throws Error 
+		{
+			DeviceList device_list;
+			Device device = null;
+			Session session;
```

Comment:
Let's move declarations as close as possible to where they are used, and `var` when possible.